safekeeper/src/wal_storage.rs

```diff
@@ -467,6 +473,39 @@ impl Storage for PhysicalStorage {
         self.update_flush_lsn();
         Ok(())
     }
+
+    fn remove_up_to(&self) -> Box<dyn Fn(XLogSegNo) -> Result<()>> {
+        let timeline_dir = self.timeline_dir.clone();
+        let wal_seg_size = self.wal_seg_size.unwrap();
+        Box::new(move |segno_up_to: XLogSegNo| {
+            remove_up_to(&timeline_dir, wal_seg_size, segno_up_to)
+        })
+    }
+}
+
+/// Remove all WAL segments in timeline_dir <= given segno.
+fn remove_up_to(timeline_dir: &Path, wal_seg_size: usize, segno_up_to: XLogSegNo) -> Result<()> {
+    for entry in fs::read_dir(&timeline_dir)? {
+        let entry = entry?;
+        let entry_path = entry.path();
+        let fname = entry_path.file_name().unwrap();
+
+        if let Some(fname_str) = fname.to_str() {
+            /* Ignore files that are not XLOG segments */
+            if !IsXLogFileName(fname_str) && !IsPartialXLogFileName(fname_str) {
+                continue;
+            }
+            let (segno, _) = XLogFromFileName(fname_str, wal_seg_size);
+            if segno <= segno_up_to {
+                remove_file(entry_path)?;
+            }
+        }
+    }
+    info!(
+        "removed WAL segments up to {}",
```

Comment:
We can also track number of files removed in the for loop, as well as min-max removed segments and print it the logs, something like this:
```
removed 3 WAL segments from 000000010000000000000001 to 000000010000000000000003
```